File: tests/parser/erlang.t/function_declaration.ml
let rec literal_atom = function | () -> `ok
let rec literal_quoted_atom = function | () -> `What.is_going:on!
let rec literal_integer = function | () -> 1
let rec literal_float = function | () -> 1.0
let rec tuple_empty = function | () -> ()
let rec tuple_nested = function | () -> ((), ())
let rec tuple_filled = function | () -> (`ok, (`error, "hello", 1), (tuple_empty ()))
let rec list_empty = function | () -> []
let rec list_nested = function | () -> [[]; [[]]]
let rec list_filled = function | () -> [`ok; [`error; "hello"; 1]; list_empty ()]
let rec list_cons = function | () -> [`a; [(`b, `c); [list_empty (); []]]]
let rec fun_args_atom = function | `ok -> `ok
let rec fun_args_quoted_atom = function | `What.is_going:on! -> `ok
let rec fun_args_integer = function | 1 -> `ok
let rec fun_args_float = function | 1.0 -> `ok
let rec fun_args_tuple_empty = function | () -> `ok
let rec fun_args_tuple_nested = function | ((), ()) -> `ok
let rec fun_args_tuple_filled = function | `ok (`error ("hello", 1)) -> `ok
let rec fun_args_list_empty = function | [] -> `ok
let rec fun_args_list_nested = function | []::([]::[])::[] -> `ok
let rec fun_args_list_filled = function | `ok::(`error::"hello"::1::[])::[] -> `ok
let rec fun_args_list_cons = function | `a::((`b `c)::(`f::[]::[])::[])::[] -> `ok
let rec fun_arg_var_in_tuple = function | a -> a
let rec fun_arg_var_in_list = function | a::[] -> a
let rec fun_arg_var = function | a -> a
let rec fun_arg_var_ignore = function | _ -> `ok
let rec fun_arg_var_ignore_in_tuple = function | (a, _) -> a
let rec fun_arg_var_ignore_in_list = function | _::b::[] -> b
let rec fun_arg_var_ignore_in_cons = function | a::_::[] -> a
let rec fun_args = function | (a, b) -> (a, b)
let rec fun_args = function | (a, b, c) -> (a, b, c)
let rec fun_args = function | (a, b, c, d) -> (a, b, c, d)
let rec fun_args = function | (a, b, c, d, e) -> (a, b, c, d, e)
let rec fun_args = function | (a, b, c, d, e, f) -> (a, b, c, d, e, f)
let rec fun_args = function | (a, b, c, d, e, f, g) -> (a, b, c, d, e, f, g)
let rec fun_args = function | (a, b, c, d, e, f, g, h) -> (a, b, c, d, e, f, g, h)
let rec fun_args = function | (a, b, c, d, e, f, g, h, i) -> (a, b, c, d, e, f, g, h, i)
let rec binding_return = function | () -> let a = 1 in a
let rec binding_and_return = function | () -> let a = 1 in a
let rec case_expr =
	function
	| a ->
			(match a with
			 | `true -> `false
			 | `true -> `false
			 | `false::[] -> `true
			 | `false::_::[] -> `true
			 | `false::b::t::[] -> `true
			 | `false -> `true)
let rec fun_ref = function | () -> fun_ref
let rec lambda = function | () -> (function | () -> `ok)
let rec lambda_with_args = function | () -> (function | a -> a)
let rec lambda_in_var = function | () -> let f = function | a -> a in f
let rec lambda_var_call =
	function | () -> let f = function | a -> a in f 1
let rec send = function | a -> Stdlib.send a a
let rec send_chain = function | a -> Stdlib.send a (Stdlib.send a a)
let rec recv = function | () -> (match recv () with | x -> x)
let rec recv_with_after = function | () -> (match recv () with | x -> x)
let rec recv_selectively =
	function
	| () ->
			(match recv () with
			 | `true -> `false
			 | `true -> `false
			 | `false::[] -> `true
			 | `false::_::[] -> `true
			 | `false::b::t::[] -> `true
			 | `false -> `true)
let rec fun_cases =
	function | 1 -> `ok | 2 -> `ok | 3 -> `ok | _ -> `false
let rec fib =
	function
	| 0 -> 0
	| 1 -> 1
	| n -> Stdlib.(+) (fib (Stdlib.(-) n 1)) (fib (Stdlib.(-) n 2))
let rec sequence =
	function
	| () ->
			let _ = print_string "hello" in
			let a = fib 2 in let _ = print_int a in `ok

File: tests/erlang.t/function_declaration.erl
-module(function_declaration).

literal_atom() -> ok.
literal_quoted_atom() -> 'What.is_going:on!'.
literal_integer() -> 1.
literal_float() -> 1.0.

tuple_empty() -> {}.
tuple_nested() -> {{}, {{}}}.
tuple_filled() -> {ok, {error, <<"hello">>, 1}, tuple_empty()}.

list_empty() -> [].
list_nested() -> [[], [[]]].
list_filled() -> [ok, [error, <<"hello">>, 1], list_empty()].

list_cons() -> [a | [{b, c} | [list_empty() | []]]].

fun_args_atom(ok) -> ok.
fun_args_quoted_atom('What.is_going:on!') -> ok.
fun_args_integer(1) -> ok.
fun_args_float(1.0) -> ok.

fun_args_tuple_empty({}) -> ok.
fun_args_tuple_nested({{}, {{}}}) -> ok.
fun_args_tuple_filled({ok, {error, <<"hello">>, 1}}) -> ok.

fun_args_list_empty([]) -> ok.
fun_args_list_nested([[], [[]]]) -> ok.
fun_args_list_filled([ok, [error, <<"hello">>, 1]]) -> ok.
fun_args_list_cons([a | [{b, c} | [f | []]]]) -> ok.

fun_arg_var_in_tuple({A}) -> A.
fun_arg_var_in_list([A]) -> A.
fun_arg_var(A) -> A.
fun_arg_var_ignore(_) -> ok.
fun_arg_var_ignore_in_tuple({A, _}) -> A.
fun_arg_var_ignore_in_list([_, B]) -> B.
fun_arg_var_ignore_in_cons([A | _]) -> A.

fun_args(A, B) -> {A, B}.
fun_args(A, B, C) -> {A, B, C}.
fun_args(A, B, C, D) -> {A, B, C, D}.
fun_args(A, B, C, D, E) -> {A, B, C, D, E}.
fun_args(A, B, C, D, E, F) -> {A, B, C, D, E, F}.
fun_args(A, B, C, D, E, F, G) -> {A, B, C, D, E, F, G}.
fun_args(A, B, C, D, E, F, G, H) -> {A, B, C, D, E, F, G, H}.
fun_args(A, B, C, D, E, F, G, H, I) -> {A, B, C, D, E, F, G, H, I}.

binding_return() -> A = 1.
binding_and_return() -> A = 1, A.

case_expr(A) ->
  case A of
    true -> false;
    {true} -> false;
    [false] -> true;
    [false|_] -> true;
    [false, B | T ] -> true;
    false -> true
  end.

fun_ref() -> fun fun_ref/0.

lambda() -> fun () -> ok end.
lambda_with_args() -> fun (A) -> A end.
lambda_in_var() -> F = fun (A) -> A end.
lambda_var_call() -> F = fun (A) -> A end, F(1).

send(A) -> A ! A.
send_chain(A) -> A ! A ! A.

recv() -> receive X -> X end.
recv_with_after() -> receive X -> X after infinity -> ok end.
recv_selectively() ->
  receive
    true -> false;
    {true} -> false;
    [false] -> true;
    [false|_] -> true;
    [false, B | T ] -> true;
    false -> true
  after
    infinity -> ok
  end.

fun_cases(1) -> ok;
fun_cases(2) -> ok;
fun_cases(3) -> ok;
fun_cases(_) -> false.

fib(0) -> 0;
fib(1) -> 1;
fib(N) -> erlang:'+'(fib(erlang:'-'(N,1)), fib(erlang:'-'(N,2))).
